src/TimeModeling/Types/judiVector.jl

```diff
@@ -98,8 +97,12 @@ function judiVector(geometry::Geometry, data::Array{T, N}) where {T, N}
     return judiVector{T, Array{T, N}}("Seismic data vector", m, n, nsrc, geometry, dataCell)
 end
 
+function judiVector(geometry::Geometry, data::Vector{T}) where {T, N}
```

Comment:
Don't see the point of this shape change.

Reply:
This is to dispatch the data in judiVector so if 1D array is input, it reshapes it to 2D array. If you set up `q = judiVector(geometry, wavelet::Vector{Float32})` then

```julia
julia> # Nonlinear modeling
       dobs = Pr*F*adjoint(Ps)*q
┌ Warning: use values(kwargs) and keys(kwargs) instead of kwargs.data and kwargs.itr
│   caller = get_exp(::Int64, ::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}) at distributed.jl:52
└ @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:52
┌ Warning: `vendor()` is deprecated, use `BLAS.get_config()` and inspect the output instead
│   caller = npyinitialize() at numpy.jl:67
└ @ PyCall ~/.julia/packages/PyCall/L0fLP/src/numpy.jl:67
ERROR: TaskFailedException

    nested task error: TaskFailedException
    
        nested task error: MethodError: no method matching SincInterpolation(::Vector{Float32}, ::StepRangeLen{Float32, Float64, Float64, Int64}, ::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64})
        Closest candidates are:
          SincInterpolation(::Matrix{T}, ::Any, ::Any) where T<:Real at ~/.julia/dev/JUDI/src/TimeModeling/Utils/auxiliaryFunctions.jl:751
        Stacktrace:
         [1] time_resample(data::Vector{Float32}, geometry_in::GeometryIC{Float32}, dt_new::Float64)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Utils/auxiliaryFunctions.jl:515
         [2] devito_interface(modelPy::PyCall.PyObject, model::Model, srcGeometry::GeometryIC{Float32}, srcData::Vector{Float32}, recGeometry::GeometryIC{Float32}, recData::Nothing, dm::Nothing, options::Options)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/python_interface.jl:7
         [3] time_modeling(model_full::Model, srcGeometry::GeometryIC{Float32}, srcData::Vector{Float32}, recGeometry::GeometryIC{Float32}, recData::Nothing, dm::Nothing, op::Char, mode::Int64, options::Options)
           @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/time_modeling_serial.jl:37
         [4] (::Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}})(::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
           @ Base ./error.jl:307
         [5] #76
           @ ./error.jl:290 [inlined]
         [6] #149
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:429 [inlined]
         [7] run_work_thunk(thunk::Distributed.var"#149#150"{Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}}, Tuple{Model, GeometryIC{Float32}, Vector{Float32}, GeometryIC{Float32}, Nothing, Nothing, Char, Int64, Options}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, print_error::Bool)
           @ Distributed /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:63
         [8] run_work_thunk
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:72 [inlined]
         [9] (::Distributed.var"#96#98"{Distributed.RemoteValue, Distributed.var"#149#150"{Base.var"#76#78"{Base.var"#76#77#79"{ExponentialBackOff, Nothing, typeof(time_modeling)}}, Tuple{Model, GeometryIC{Float32}, Vector{Float32}, GeometryIC{Float32}, Nothing, Nothing, Char, Int64, Options}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}})()
           @ Distributed ./task.jl:423
        Stacktrace:
          [1] #remotecall_fetch#154
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:455 [inlined]
          [2] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
          [3] #remotecall_fetch#158
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
          [4] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
          [5] local_reduce!(my_future::Distributed.Future, other_future::Distributed.Future)
            @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:192
          [6] #149
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:429 [inlined]
          [7] run_work_thunk(thunk::Distributed.var"#149#150"{typeof(JUDI.local_reduce!), Tuple{Distributed.Future, Distributed.Future}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, print_error::Bool)
            @ Distributed /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/process_messages.jl:63
          [8] #remotecall_fetch#154
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
          [9] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
         [10] #remotecall_fetch#158
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [11] remotecall_fetch
            @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [12] (::JUDI.var"#190#191"{Vector{Distributed.Future}, Int64})()
            @ JUDI ./task.jl:423
        Stacktrace:
         [1] #remotecall_fetch#154
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:455 [inlined]
         [2] remotecall_fetch
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:454 [inlined]
         [3] #remotecall_fetch#158
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [4] remotecall_fetch
           @ /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Distributed/src/remotecall.jl:496 [inlined]
         [5] (::JUDI.var"#190#191"{Vector{Distributed.Future}, Int64})()
           @ JUDI ./task.jl:423
    Stacktrace:
     [1] sync_end(c::Channel{Any})
       @ Base ./task.jl:381
     [2] macro expansion
       @ ./task.jl:400 [inlined]
     [3] reduce_level!(futures::Vector{Distributed.Future}, nleaf::Int64)
       @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:208
     [4] reduce!(futures::Vector{Distributed.Future})
       @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:242
     [5] (::JUDI.var"#187#189"{Vector{Any}, Int64, Vector{Distributed.Future}})()
       @ JUDI ./task.jl:423

...and 1 more exception.

Stacktrace:
  [1] sync_end(c::Channel{Any})
    @ Base ./task.jl:381
  [2] macro expansion
    @ ./task.jl:400 [inlined]
  [3] task_distributed(::Function, ::Distributed.WorkerPool, ::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:132
  [4] task_distributed(::Function, ::Distributed.WorkerPool, ::Model, ::Vararg{Any})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:125
  [5] time_modeling(::Model, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:4
  [6] time_modeling
    @ ~/.julia/dev/JUDI/src/TimeModeling/Modeling/distributed.jl:4 [inlined]
  [7] (::JUDI.var"#221#223"{Options, Info, Model, GeometryIC{Float32}, GeometryIC{Float32}, UnitRange{Int64}})(src::judiVector{Float32, Vector{Float32}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/LinearOperators/judiPDEfull.jl:52
  [8] *(A::judiPDEfull{Float32, Float32}, v::judiVector{Float32, Vector{Float32}})
    @ JUDI ~/.julia/dev/JUDI/src/TimeModeling/LinearOperators/judiPDEfull.jl:104
  [9] afoldl
    @ ./operators.jl:612 [inlined]
 [10] *(a::judiProjection{Float32, Float32}, b::judiModeling{Float32, Float32}, c::judiProjection{Float32, Float32}, xs::judiVector{Float32, Vector{Float32}})
    @ Base ./operators.jl:655
 [11] top-level scope
    @ REPL[40]:2
```

Reply:
So instead of just fixing `SincInterpolation` to work on vector you enforce painful reshape everywhere?

Reply:
Let me put it this way, if `wavelet::Vector{T}`:

`q1 = judiVector(geometry, wavelet)`

and

`q2 = judiVector(geometry, reshape(wavelet, :, 1))`

should be the same source function. The problem is now they are NOT. So this reshape should be enforced so that `q.data` is always a `Matrix` instead of a `Vector`.

Reply:
No! This reshape is enforced "by you" and the force reshape you introduce then you change the test to pretend that change isn't there and messing things up.
Both `q1` and `q2` are valid and no it should not be the same since it's a different input, there is zero reason to enforce they the same.

Reply:
I would argue one more time on this:

if we want to save q1 and q2 as SEGY files, we have to dispatch there again (because `q1.data::Vector` and `q2.data::Matrix`). Now this question comes again: should `segy(q1)` be the same as `segy(q2)`? There might be a lot of problems associated with this so I find it quite weird to make things like this, i.e. source functions not having a standardized form.

We might experience other issues later if we further develop some new features. We will always face the problem that we are uncertain about the type of `q.data`. Why don't we enforce it to be a `Matrix` which can really save a lot of dispatches in the future, and things could be in standard forms?

This is my personal opinion but feel free to convince me if this isn't a good practice.

Reply:
You are enforcing implemetation details onto the user which is wrong. The dispatch is *made* for different types to to prevent it, that's why it's called multiple dispatch, mot "single dispatch or we error". 

`Vector` and `Matrix` are `::Array{T, N}` so your dispatch is fixed, done. This whole PR is literally two line change
- FIxe Sinc to accept Vector (which is what dispatch is for)
- Add the `[1]` for sizxe one
And that's it. Everything else is convoluted way to bypass a "single type enforcing" rather than a "diaptch".

Reply:
OK then the problem is that `judiVector->segy->judiVector` isn't identity operation. We don't want that

Reply:
> But enforcing SEGY limitation into judiVector is backward.

I would argue this also entails the forward process -- if `judiVector(geometry, data::Vector)` and `judiVector(geometry, data::Matrix)` end up to have the same SEGY file, then you have no way to ensure backward compatibility

Reply:
That's not what backward compatibility means.

I only meant backward that it's the wrong way to design it.

Reply:
OK then we should do:

We accept `data::Vector` for `judiVector` but when we load SegyIO file, we always make `data::Matrix` for `judiVector`. Does that sound good?

Reply:
Why? The current version reads it as a Vector what's the issue with it.